### PR TITLE
feat: Framework splitting

### PR DESCRIPTION
**Pull request by MikhailKardash**
## Ticket
[MD-272]

## Description
Makes pytorch-ngc our default image. Sets tensorflow-ngc in tests when required.
Moves legacy PyTorchTrial launcher from `horovod` to `torch.distributed.launch`.
Lots of docs updates.
Update master restart test to include a configurable timeout. In this case, we have to pull the pytorch image, so we timeout. This will be fixed in a later `environments` PR, but this has to be merged first.
Also splits our tests a bit better at the directory level to make a clearer distinction between PyTorch and TensorFlow.

## Test Plan
Unit tests pass, experiments default to pytorch-ngc images.
Verify that Tensorboard launches from ngc-based image experiments.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[MD-272]: https://hpe-aiatscale.atlassian.net/browse/MD-272?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by keita-determined — APPROVED** [review]

**Review by djanicekpach — APPROVED** [review]
CI config changes seemed reasonable to me.

File: docs/manage/troubleshooting.rst [hunk 1]
```diff
@@ -4,6 +4,18 @@
...
+      TensorFlow users must configure their environment image in their experiment configuration file
+      before submitting an experiment.
+
+      .. code:: bash
+
+         environment:
+            image:
+               cpu: determinedai/tensorflow-ngc-dev:f17151a
+               gpu: determinedai/tensorflow-ngc-dev:f17151a
+
```
Comment (tara-det-ai) [inline docs/manage/troubleshooting.rst:18]: .. important::

      TensorFlow users must configure the environment image in their :ref:`experiment configuration
         <experiment-config-reference>` file
      before submitting an experiment.

      .. code:: bash
         environment:
            image:
               cpu: determinedai/tensorflow-ngc-dev:f17151a
               gpu: determinedai/tensorflow-ngc-dev:f17151a

File: docs/release-notes/framework-splitting.rst [hunk 2]
```diff
@@ -0,0 +1,10 @@
+:orphan:
+
+**Breaking Change**
+
+-  Default images are being changed to support only PyTorch by default. TensorFlow users must
+   configure their experiments to target our non-default TensorFlow images. Details on this process
+   can be found at :ref:`set-environment-images`
+
+-  Our new default images are built off of Nvidia NGC images. We select an NGC version, however
+   users can build their own images from an NGC version of their choice. See :ref:`ngc-version`
```
Comment (tara-det-ai) [inline docs/release-notes/framework-splitting.rst:10]: :orphan:

**Breaking Changes**

-  Images: The default environment includes images that support PyTorch. TensorFlow users must configure their experiments to target our non-default TensorFlow images. Details on this process can be found at :ref:`set-environment-images`.

-  Images: Our new default images are based on Nvidia NGC. While we provide a recommended NGC version, users have the flexibility to build their own images using any NGC version that meets their specific requirements. For more information, visit :ref:`ngc-version`.

**Review by tara-det-ai — APPROVED** [review]
suggested edits

File: harness/tests/experiment/keras/test_tf_keras_trial.py [hunk 3]
```diff
@@ -593,3 +593,33 @@ def make_workloads_with_metrics(metrics_storage: List) -> workload.Stream:
 
     assert len(trial_B_metrics) == len(trial_C_metrics) == 5
     assert trial_B_metrics == trial_C_metrics
+
+
+@pytest.mark.tensorflow
+def test_keras_from_config() -> None:
+    from tests.experiment.fixtures import tf_keras_one_var_model
```
Comment (highvelcty) [inline harness/tests/experiment/keras/test_tf_keras_trial.py:600]: I suggest moving this to the top of the file or adding a comment about why it is here.

**Review by highvelcty — APPROVED** [review]
I am new to determined and new to this code set, but the changes look good to me! I made one minor comment about an import within a test method instead of at the module level. Feel free to take it or leave it :). Thanks for the changes!

**Review by gt2345 — APPROVED** [review]
Confirming that experiment image gets updated

File: harness/tests/experiment/fixtures/ancient-checkpoints/0.17.6-keras/metadata.json [hunk 4]
```diff
@@ -39,8 +39,8 @@
       },
       "force_pull_image": false,
       "image": {
-        "cpu": "determinedai/pytorch-tensorflow-cpu-dev:8b3bea3",
-        "cuda": "determinedai/pytorch-tensorflow-cuda-dev:8b3bea3",
+        "cpu": "determinedai/tensorflow-ngc-dev:8b3bea3",
+        "cuda": "determinedai/tensorflow-ngc-dev:8b3bea3\"",
```
Comment (azhou-determined) [inline harness/tests/experiment/fixtures/ancient-checkpoints/0.17.6-keras/metadata.json:43]: typo?

**Comment by azhou-determined** [comment]
do we install tensorboard in the pytorch-only images? IIRC, launching a tensorboard from an experiment will default to the image defined in an experiment. 

another thing to test would be if you launch a tensorboard with multiple experiments, where one has pytorch-only and another is tf-only.

after this change, what image do we expect to be used to launch NTSC?

**Comment by azhou-determined** [comment]
general question: is there a doc somewhere that catalogues all the different images we offer and their intended use?

**Comment by MikhailKardash** [comment]
> do we install tensorboard in the pytorch-only images? IIRC, launching a tensorboard from an experiment will default to the image defined in an experiment.
Yes we do, that's done in environments: https://github.com/determined-ai/environments/blob/8b3bea3b81bd3934ad885bdf159320f11f6ea1ba/Dockerfile-pytorch-ngc#L21
> another thing to test would be if you launch a tensorboard with multiple experiments, where one has pytorch-only and another is tf-only.
I'll verify this and then add it to testing instructions.
> after this change, what image do we expect to be used to launch NTSC?
After this change, we expect to launch in `determinedai/pytorch-ngc-dev:<HASH>` images. These will be retagged by release-party to be `determinedai/pytorch-ngc:<VERSION>`

**Comment by MikhailKardash** [comment]
> general question: is there a doc somewhere that catalogues all the different images we offer and their intended use?

We have this page in our docs: https://docs.determined.ai/latest/model-dev-guide/prepare-container/set-environment-images.html#set-environment-images
Which is updated in this PR: `docs/model-dev-guide/prepare-container/set-environment-images.rst`

**Comment by azhou-determined** [comment]
> > do we install tensorboard in the pytorch-only images? IIRC, launching a tensorboard from an experiment will default to the image defined in an experiment.
> 
> Yes we do, that's done in environments: https://github.com/determined-ai/environments/blob/8b3bea3b81bd3934ad885bdf159320f11f6ea1ba/Dockerfile-pytorch-ngc#L21


if we only install `torch-tb-profiler` in torch images, and `tensorflow-plugin-profile` in tf images, then the case when someone compares multiple experiments where one profiles with torch and the other with TF profiler, seems this might cause issues?

**Comment by MikhailKardash** [comment]
> seems like that environments image only installs the profilers? where does actual `tensorboard` get installed, is it in the base image maybe?

`torch-tb-profiler` installs `tensorboard` as a dependency. See: https://github.com/pytorch/kineto/blob/84d95b8232167674eee17c11c2198276f4a6482c/tb_plugin/setup.py#L29
They version pin `tensorboard` on their end anyways.

> also, if we only install `torch-tb-profiler` in torch images, and `tensorflow-plugin-profile` in tf images, then the case when someone compares multiple experiments where one profiles with torch and the other with TF profiler, seems this might cause issues?

They do all load in tensorboard right now. See the following image, where trial 3 is a pytorch mnist trial while trials 4-6 are iris tfkeras trials. Note that the tensorboard inherits the image used in the trial it is associated with. I tried with `ngc-tensorflow-dev`, `ngc-pytorch-dev`, and even `pytorch-tensorflow` and they all work.
image

**Review by azhou-determined — APPROVED** [review]
nice
